File: day10_part1_2021.rb

def syntax_error_score(file)
  scores = { ')' => 3, ']' => 57, '}' => 1197, '>' => 25137 }
  pairs = { '(' => ')', '[' => ']', '{' => '}', '<' => '>' }
  total_score = 0

  File.readlines(file).each do |line|
    stack = []
    line.strip.each_char do |char|
      if pairs.keys.include?(char)
        stack.push(char)
      else
        if char == pairs[stack.last]
          stack.pop
        else
          total_score += scores[char]
          break
        end
      end
    end
  end

  total_score
end

puts syntax_error_score('input.txt')

File: day10_part1_2021.R

input <- readLines("input.txt")

illegal_chars <- c(")", "]", "}", ">")
illegal_points <- c(3, 57, 1197, 25137)

total_score <- 0

for (line in input) {
  if (grepl("[\\[\\{\\(\\<]", line)) {
    stack <- c()
    for (char in strsplit(line, "")[[1]]) {
      if (char %in% c("(", "[", "{", "<")) {
        stack <- c(stack, char)
      } else if (char %in% c(")", "]", "}", ">")) {
        if (length(stack) == 0 || (char == ")" && stack[length(stack)] != "(") ||
            (char == "]" && stack[length(stack)] != "[") ||
            (char == "}" && stack[length(stack)] != "{") ||
            (char == ">" && stack[length(stack)] != "<")) {
          total_score <- total_score + illegal_points[which(illegal_chars == char)]
          break
        } else {
          stack <- stack[-length(stack)]
        }
      }
    }
  }
}

print(total_score)
